pkg/api/customization/setting/setting.go

```diff
@@ -36,13 +37,20 @@ func Formatter(apiContext *types.APIContext, resource *types.RawResource) {
 
 func Validator(request *types.APIContext, schema *types.Schema, data map[string]interface{}) error {
 	var setting v3client.Setting
-	if err := access.ByID(request, request.Version, v3client.SettingType, request.ID, &setting); err != nil {
+
+	// request.ID is taken from the request request url, it is possible that the request url does not contain the id
+	id := request.ID
+	if name, ok := data["name"].(string); ok && id == "" {
+		id = name
+	}
+
+	if err := access.ByID(request, request.Version, v3client.SettingType, id, &setting); err != nil && !strings.Contains(err.Error(), "NotFound") {
 		return err
 	}
 	if setting.Source == "env" {
```

Comment:
I think you can change to
```
if err := access.ByID(request, request.Version, v3client.SettingType, request.ID, &setting); err == nil {
if setting.Source == "env" { // it means that user is edit a readonly resource, we don't allow it.
return httperror.NewAPIError(httperror.MethodNotAllowed, fmt.Sprintf("%s is readOnly because its value is from environment variable", request.ID))
}
}
```

Reply:
We also need to check `slice.ContainsString(readOnlySettings, setting.ID)`.

Reply:
@orangedeng I dont understand what you're suggesting here. Did you possible mean "err != nil"?

Your suggestion doesnt really explain what is wrong with @rmweir 's